### PR TITLE
Add "reserved" attribute for an enum variant, allowing derive(FromBits) even if unfilled

### LINKED ISSUE #24: Add reserved enum variants, allowing derive(FromBits) even if unfilled
See #23.

Also, just to document it, this is mostly about structs with such reserved enum fields:
```rust
StructUnfilled::try_from(...);
```
would just return `Err(num)`, even though it might just be one wrong field.

For enums alone, it wouldn't matter, since:
```rust
enum Unfilled {
  A, B, C, /*some missing*/
}
match Unfilled::try_from(..) {
  Ok(A) | Ok(B) | Ok(C) => /*...*/
  Err(missing_value) => /*...*/
}
```
is doing kinda the same thing as:
```rust
enum Unfilled {
A, B, C, CatchAll(u4), //or `Reserved,`
}
match Unfilled::from(..) {
  A | B | C => /*...*/
  //or `Reserved =>`
  CatchAll(missing_value) => /*...*/
}
```

Note: serde uses `#[serde(other)]` to do this, `num_enum` uses `#[default]` or `#[num_enum(default)]`.
We'll use `#[fallback]`, which tells you what it does. This is fine until someone uses both our macro and some other macro with the exact same attribute. When that happends, of course I'm open to provide a scoped version too, i.e. `#[bilge(fallback)]`.

### PR DESCRIPTION
**Pull request by pickx**
Resolves #24.
Currently, enums which don't fill their bitfield don't allow deriving `FromBits`. This PR allows adding a `#[reserved]` attribute to one (and only one) variant of such enums:
```rust
#[bitsize(3)]
#[derive(FromBits)]
enum Opcode {
    Add = 0b001,
    Sub = 0b010,
    Mul = 0b011,
    #[reserved]
    Reserved,
}
```

This will then generate a `_ => Self::Reserved` arm in the `From<int>` implementation. This is is a followup from the reddit thread, where I realized this feature could be useful to me. 

Please note that **this is a WIP**: for example, this trips the generated `assert!(#type_name::FILLED)` check, and so the generated code fails to compile. I wanted to see if this PR aligns with what you want this crate to look like, before attempting to change that assert.

I'm open to any changes or suggestions. Thanks for making this crate.

### CODE REVIEW

File: bilge-impl/src/from_bits.rs
Comment (hecatia-elegua) [inline bilge-impl/src/from_bits.rs:51]: Hmm, I don't think it's that easy. Just now I _thought_ we just need to note that `FILLED` is not respecting `#[reserved]` fields, but we actually need it to respect them since structs need to know if their fields are filled.
So, you have to just get the reserved variants here as well:
https://github.com/hecatia-elegua/bilge/blob/f59d6600f59195a8240752f0ed32823167a4e4d9/bilge-impl/src/bitsize.rs#L238

**Review by hecatia-elegua — CHANGES_REQUESTED** [review]
Hey, thank you a ton for your contribution!
I love some of the readability improvements, like `.any(is_non_exhaustive_attribute)`.

For the top-level `from_bits`/`try_from_bits`, the rule is kinda "anything only related to this derive should live here", which is why I'm not sure about `DeriveImpl`. But then again, I also did the same thing by passing `try_from` in, so... maybe we should just keep the function-skeleton, like we did for `parse` and `analyze(_enum)`? Please let us discuss here first since I don't wanna pointlessly have you shuffle code around until I esoterically approve lol.

This is similar to what num-enum does, `#[num_enum(default)]` or just `#[default]`. Since it's not really a default value, I think your naming is better. We could bike-shed if we want to call it `#[other(s?)]`, since I could imagine an enum field which fills up all other values with a special variant or sth. Or we just add that later. Same for `#[bilge(reserved)]` or `#[bilge(other)]`, which is just about namespacing issues with multiple attribute macros.

Some notes on how to do `FILLED` below:

File: bilge-impl/src/shared.rs
Comment (pickx) [reply bilge-impl/src/shared.rs:43]: yes, it should. thinking about it again, there's no reason to allow `#[reserved]` with `TryFromBits`.

**Comment by pickx** [comment]
Hey, thanks for the quick and thorough response. 
I do think it makes sense to re-introduce the separation between the two derives. As you have pointed out, `#[reserved]` shouldn't be allowed with `TryFromBits` (and it should abort to inform the user about this). Thus the `TryFromBits` implementations doesn't need to know about this attribute and we can keep the whole reserved variant logic inside `from_bits.rs`. Following this line of thought, maybe `analyze_enum_derive()` should be split into two versions and not take `try_from`. I'm not sure.

Regarding the name, `reserved` does seem like it could clash with some other crate's attribute. I chose that name because it already has a similar meaning in bilge's structs. Again, not sure. Your call.

As for `FILLED` - generating `assert!(true)` would work. Another, possibly hackier option could be to also generate a `const RESERVED: Option<Self>` in the impl, which would be the reserved variant or `None` if no `#[reserved]`. Then the assert would be `assert!(#type_name::FILLED || #type_name::RESERVED.is_some())`. This has the benefit (?) of also letting the enum know about its reserved variant after generation, which I guess might be useful somehow.

**Comment by hecatia-elegua** [comment]
Actually, `#[reserved]` is `from_bits` `enum`-only, so, without looking at the code again, can we just put it in the `from_bits.rs` `analyze_enum` code path? I also agree with really going for the split. Back then I also regretted a bit that I handled it with a bool, I should've just called multiple smaller functions in `shared.rs` instead. If you're up to it, do it, otherwise I can do it myself after this PR.

I like `fallback`! Use that, it sounds super clear. I can see `reserved` being useful for something else.

Let's leave out `RESERVED` but I'll keep it in mind.

**Comment by pickx** [comment]
> Actually, `#[reserved]` is `from_bits` `enum`-only, so, without looking at the code again, can we just put it in the `from_bits.rs` `analyze_enum` code path? 
yep, I tried doing as much as possible, though I did made `struct`s also check for a fallback (and so kept it in the derive analysis part), so they can fail.
I hope the changes I made are not radical, obviously I'm always up for suggestions.

File: bilge-impl/src/shared.rs
Comment (hecatia-elegua) [inline bilge-impl/src/shared.rs:184]: Nice, that's the "mathematic" version of `max_variants_count`. Let's decide for one of these:
`max_variants_count = 1u128 << bitsize`, `val >= max`
`max_variants_count = 2u128^bitsize - 1`, `val > max`

bitsize = 2, val = 3
`1 << 2 == 4`, `3 >= 4 == false`
`2^2 -1 == 3`, `3 > 3 == false`

File: bilge-impl/src/try_from_bits.rs
Comment (hecatia-elegua) [inline bilge-impl/src/try_from_bits.rs]: That's a really good idea, again :)
(removing low level enum handling from this high level code)

**Review by hecatia-elegua — APPROVED** [review]
I would've put the new `EnumVariantValueAssigner` change in a different PR, but I know the hassle :)
It's a nice change, let's keep it here.
Some more comments, but I'll just merge this now since it's pedantic stuff:

File: bilge-impl/src/shared.rs
Comment (pickx) [reply bilge-impl/src/shared.rs:184]: it is faster, but `1u128 << bitsize` would panic for a `bitsize` of 128, while the `saturating_pow` version would not.

Comment (pickx) [reply bilge-impl/src/shared.rs:185]: oh! I didn't notice. so yeah, no reason not to use a bitshift.
are you okay with: `(1u128 << bitsize) - 1` and `val > max`?

Comment (hecatia-elegua) [reply bilge-impl/src/shared.rs:185]: yes

Comment (pickx) [reply bilge-impl/src/shared.rs]: done. thanks.

**Comment by hecatia-elegua** [comment]
Thanks!

**Comment by pickx** [comment]
> @pickx btw did you want your name to be slapped onto these commits? Because that won't link to your github profile now, which is not so nice _for_ you.
> 
> There's also this, which you can do with every commit on github, by adding `.patch` to the end of the link: https://github.com/hecatia-elegua/bilge/commit/bd2e319d22b71c814a77d26df094f6c38bafefcd.patch which leaks your e-mail (not sure if that matters to you)

yeah... uh, that's not what I expected to happen. I now see that my `.gitconfig` is to blame. 
sorry, but is it possible to change this to use my github account?

**Comment by hecatia-elegua** [comment]
Github is kinda dumb at stuff like this.
I've noticed some of your commits last year have the same problem, not sure if you want to fix all those too.

You'd have to use interactive rebase on all these commits locally and change your author information, then force push.
Then I'd have to do the same on the main branch. That should be it?

**Comment by pickx** [comment]
@hecatia-elegua okay, I seem to have done it. for reference, this is an easy way to do it while preserving commit dates https://stackoverflow.com/a/70190223

**Comment by hecatia-elegua** [comment]
@pickx I think it didn't work. I've added two links to the previous message, that way should work.

**Comment by pickx** [comment]
> Oh wait you just have to force push or sth? Or does github not register that you changed it? Probably because contributors could then destroy old PR histories... hm

yeah, that seems to be it. while the branch in my fork changed, the one in this pull request didn't. my assumption is that you also have to do this on your own repo.

**Comment by hecatia-elegua** [comment]
Well, I've changed it on main, so it's only left in here and your old repos.
If you want we can delete our comments here to make it less obvious, but I'm guessing you only wanted the github account linkage, not privacy? Although you use protonmail hmm...
btw. you can message me on discord, too, if you have that (hecatia-elegua#1559) 👍

**Comment by pickx** [comment]
sent you a friend request (pike#8242). 
thanks again, that was my first open source commit (could you tell? lol).
I'm fine with keeping the PR page way it is. not a big deal.

**Pull request merged**
